File: Operators/update_round_robin.m
function [output1,output2] = update_round_robin(varargin)
% Conduct update via round-robin tournament.

mode = varargin{end};
switch mode
    case 'execute'
        Solution = varargin{1};
        Problem  = varargin{2};

        Fitness  = Solution.fits;
        K   = min(10,Problem.N-1);
        ind = 1:length(Solution);
        win = zeros(length(Solution),1);
        for i = 1:length(Solution)
            win(i) = sum(Fitness(i)<=Fitness(randperm(end,K)));
        end
        [~,rank] = sort(win,'descend');
        ind = ind(rank(1:Problem.N));
        output1 = Solution(ind);

    case 'parameter'
        % n/a

    case 'behavior'
        output1 = {'';''}; 
end

if ~exist('output1','var')
    output1 = [];
end
if ~exist('output2','var')
    output2 = [];
end
end

File: AutoOptLib for Python/components/update_round_robin.py
"""Python translation of update_round_robin."""

from __future__ import annotations

import numpy as np

from ._utils import ensure_rng, extract_fits, flex_get, solution_as_list


def update_round_robin(*args):
    mode = args[-1]
    if mode == "execute":
        solution = args[0]
        problem = args[1] if len(args) > 1 else None
        aux = args[3] if len(args) > 3 else None
        rng = ensure_rng(aux)

        sol_list = solution_as_list(solution)
        fitness = extract_fits(solution)
        pop_size = len(sol_list)
        if pop_size == 0:
            return [], None

        n = int(flex_get(problem, "N", pop_size))
        n = min(n, pop_size)
        k = min(10, n - 1)
        if k <= 0:
            selected = np.argsort(fitness)[:n]
            return [sol_list[i] for i in selected], None

        win = np.zeros(pop_size, dtype=int)
        for i in range(pop_size):
            opponents = rng.choice(pop_size, size=k, replace=False)
            win[i] = int(np.sum(fitness[i] <= fitness[opponents]))

        rank = np.argsort(win)[::-1]
        chosen = rank[:n]
        return [sol_list[i] for i in chosen], None

    if mode == "parameter":
        return None, None

    if mode == "behavior":
        return ["", ""], None

    raise ValueError(f"Unsupported mode: {mode}")
